File: app/lib/models/report.dart
import 'package:flutter/foundation.dart';

/// One numbered line of the Daily Maintenance Report.
@immutable
class DmrLine {
  const DmrLine({
    required this.number,
    required this.label,
    required this.key,
    required this.derived,
    this.value,
    this.isDecimal = false,
  });

  final int number;
  final String label;

  /// The field name on the wire — what an edit sends back.
  final String key;

  /// False when nothing in the system observes it, so a person enters it.
  final bool derived;
  final double? value;
  final bool isDecimal;

  bool get isBlank => value == null;

  /// How the depot writes it: counts whole, quantities to one decimal, and a
  /// dash where nothing has been recorded — never a misleading zero.
  String get display {
    final v = value;
    if (v == null) return '—';
    return isDecimal ? v.toStringAsFixed(1) : v.round().toString();
  }

  factory DmrLine.fromJson(Map<String, dynamic> json) => DmrLine(
        number: json['number'] as int,
        label: json['label'] as String,
        key: json['key'] as String,
        derived: json['derived'] as bool? ?? true,
        value: (json['value'] as num?)?.toDouble(),
        isDecimal: json['is_decimal'] as bool? ?? false,
      );
}

/// One day of the report: derived lines computed, entered lines as stored.
@immutable
class DmrDay {
  const DmrDay({
    required this.siteCode,
    required this.reportDate,
    required this.lines,
    this.notes = '',
    this.isSnapshot = false,
    this.generatedAt,
  });

  final String siteCode;
  final String reportDate;
  final List<DmrLine> lines;
  final String notes;

  /// True once the day is frozen; until then the derived lines recompute.
  final bool isSnapshot;
  final String? generatedAt;

  static const empty = DmrDay(
    siteCode: '',
    reportDate: '',
    lines: <DmrLine>[],
  );

  List<DmrLine> get derived => lines.where((l) => l.derived).toList();

  List<DmrLine> get entered => lines.where((l) => !l.derived).toList();

  /// Entered lines still waiting on somebody — what the form nudges about.
  int get unanswered => entered.where((l) => l.isBlank).length;

  factory DmrDay.fromJson(Map<String, dynamic> json) => DmrDay(
        siteCode: json['site_code'] as String? ?? '',
        reportDate: json['report_date'] as String,
        lines: <DmrLine>[
          for (final l in (json['lines'] as List<dynamic>? ?? <dynamic>[]))
            DmrLine.fromJson(l as Map<String, dynamic>),
        ],
        notes: json['notes'] as String? ?? '',
        isSnapshot: json['is_snapshot'] as bool? ?? false,
        generatedAt: json['generated_at'] as String?,
      );
}

/// The month grid: parameters down the page, one column per day.
@immutable
class DmrMonth {
  const DmrMonth({
    required this.siteCode,
    required this.month,
    required this.dates,
    required this.lines,
    required this.values,
  });

  final String siteCode;
  final String month;
  final List<String> dates;
  final List<DmrLine> lines;

  /// line key -> one value per date, in [dates] order.
  final Map<String, List<double?>> values;

  static const empty = DmrMonth(
    siteCode: '',
    month: '',
    dates: <String>[],
    lines: <DmrLine>[],
    values: <String, List<double?>>{},
  );

  List<double?> valuesFor(String key) => values[key] ?? const <double?>[];

  factory DmrMonth.fromJson(Map<String, dynamic> json) => DmrMonth(
        siteCode: json['site_code'] as String? ?? '',
        month: json['month'] as String? ?? '',
        dates: List<String>.from(json['dates'] as List<dynamic>? ?? <dynamic>[]),
        lines: <DmrLine>[
          for (final l in (json['lines'] as List<dynamic>? ?? <dynamic>[]))
            DmrLine.fromJson(l as Map<String, dynamic>),
        ],
        values: <String, List<double?>>{
          for (final e in (json['values'] as Map<String, dynamic>? ??
                  <String, dynamic>{})
              .entries)
            e.key: <double?>[
              for (final v in (e.value as List<dynamic>))
                (v as num?)?.toDouble(),
            ],
        },
      );
}

/// How the report groups a defect or a breakdown.
enum DefectCategory {
  mechanical('mechanical', 'Mechanical'),
  electrical('electrical', 'Electrical'),
  body('body', 'Body'),
  ac('ac', 'AC'),
  its('its', 'ITS'),
  tyre('tyre', 'Tyre'),
  other('other', 'Other');

  const DefectCategory(this.wireName, this.label);

  final String wireName;
  final String label;

  static DefectCategory fromWire(String? value) =>
      DefectCategory.values.firstWhere(
        (c) => c.wireName == value,
        orElse: () => DefectCategory.other,
      );
}

/// A bus off the road, from the day it went down to the day it ran again.
@immutable
class OffRoadCase {
  const OffRoadCase({
    required this.id,
    required this.vehicleId,
    required this.registrationNo,
    required this.issue,
    required this.category,
    required this.offRoadSince,
    this.model = '',
    this.odometerKm,
    this.actionTaken,
    this.expectedDays,
    this.expectedReadyOn,
    this.returnedOn,
    this.spareParts,
    this.remarks,
    this.awaitingVendor = false,
    this.daysDown = 0,
    this.isHeld = false,
  });

  final String id;
  final String vehicleId;
  final String registrationNo;
  final String model;
  final int? odometerKm;
  final String issue;
  final String? actionTaken;
  final DefectCategory category;
  final String offRoadSince;
  final int? expectedDays;
  final String? expectedReadyOn;

  /// Null while the bus is still down.
  final String? returnedOn;
  final String? spareParts;
  final String? remarks;

  /// Waiting on someone else — EKA, Octillion, JTAC. Why a day slipped.
  final bool awaitingVendor;

  /// How long it has been down as at the date asked for.
  final int daysDown;

  /// Down longer than the report's threshold.
  final bool isHeld;

  bool get isOpen => returnedOn == null;

  /// True once the promised return date has passed and it is still down.
  bool overdueAgainst(String today) =>
      isOpen &&
      expectedReadyOn != null &&
      expectedReadyOn!.compareTo(today) < 0;

  factory OffRoadCase.fromJson(Map<String, dynamic> json) => OffRoadCase(
        id: json['id'] as String,
        vehicleId: json['vehicle_id'] as String? ?? '',
        registrationNo: json['registration_no'] as String? ?? '',
        model: json['model'] as String? ?? '',
        odometerKm: (json['odometer_km'] as num?)?.round(),
        issue: json['issue'] as String? ?? '',
        actionTaken: json['action_taken'] as String?,
        category: DefectCategory.fromWire(json['category'] as String?),
        offRoadSince: json['off_road_since'] as String,
        expectedDays: json['expected_days'] as int?,
        expectedReadyOn: json['expected_ready_on'] as String?,
        returnedOn: json['returned_on'] as String?,
        spareParts: json['spare_parts_required'] as String?,
        remarks: json['remarks'] as String?,
        awaitingVendor: json['awaiting_vendor'] as bool? ?? false,
        daysDown: json['days_down'] as int? ?? 0,
        isHeld: json['is_held'] as bool? ?? false,
      );
}

/// Annexure-V: the root-cause follow-up to one breakdown.
@immutable
class Investigation {
  const Investigation({
    required this.entryId,
    required this.registrationNo,
    required this.entryDate,
    this.model = '',
    this.odometerKm,
    this.driverId,
    this.defectType = '',
    this.breakdownReason = '',
    this.location,
    this.breakdownTime,
    this.mechanicReportedTime,
    this.attendedTime,
    this.lossKm,
    this.attendedDetails,
    this.findings,
    this.lastPmOn,
    this.lastPmFindings,
    this.relatedComplaints,
    this.investigationAction,
    this.isComplete = false,
    this.updatedBy = '',
  });

  final String entryId;
  final String registrationNo;
  final String model;
  final int? odometerKm;
  final String? driverId;
  final String defectType;
  final String breakdownReason;
  final String? location;
  final String? breakdownTime;
  final String? mechanicReportedTime;
  final String? attendedTime;
  final double? lossKm;
  final String? attendedDetails;
  final String entryDate;

  /// The investigation itself — what was found and what was done about it.
  final String? findings;

  /// Answered from the inspection history rather than looked up by hand.
  final String? lastPmOn;
  final String? lastPmFindings;
  final String? relatedComplaints;
  final String? investigationAction;

  /// A finding and an action. Without both it is a placeholder.
  final bool isComplete;
  final String updatedBy;

  factory Investigation.fromJson(Map<String, dynamic> json) => Investigation(
        entryId: json['entry_id'] as String,
        registrationNo: json['registration_no'] as String? ?? '',
        model: json['model'] as String? ?? '',
        odometerKm: (json['odometer_km'] as num?)?.round(),
        driverId: json['driver_id'] as String?,
        defectType: json['defect_type'] as String? ?? '',
        breakdownReason: json['breakdown_reason'] as String? ?? '',
        location: json['location'] as String?,
        breakdownTime: json['breakdown_time'] as String?,
        mechanicReportedTime: json['mechanic_reported_time'] as String?,
        attendedTime: json['attended_time'] as String?,
        lossKm: (json['loss_km'] as num?)?.toDouble(),
        attendedDetails: json['attended_details'] as String?,
        entryDate: json['entry_date'] as String,
        findings: json['findings'] as String?,
        lastPmOn: json['last_pm_on'] as String?,
        lastPmFindings: json['last_pm_findings'] as String?,
        relatedComplaints: json['related_complaints'] as String?,
        investigationAction: json['investigation_action'] as String?,
        isComplete: json['is_complete'] as bool? ?? false,
        updatedBy: json['updated_by'] as String? ?? '',
      );
}

File: backend/app/schemas/report.py
from __future__ import annotations

from datetime import date as date_t

from pydantic import BaseModel, Field

from app.models.enums import DefectCategory
from app.schemas.common import HHMM, DecimalOut, ISTDateTime
from app.services.control_charts import CellMark, ChartKind


class DmrLine(BaseModel):
    """One numbered line of the report."""

    number: int
    label: str
    key: str
    #: False when nothing observes it, so a person enters it.
    derived: bool
    #: A JSON number, never a string — the client casts these to num.
    value: DecimalOut = None
    #: Rendered with a decimal rather than as a count.
    is_decimal: bool = False


class DmrDayOut(BaseModel):
    site_code: str
    report_date: date_t
    lines: list[DmrLine]
    notes: str = ""
    #: True once the day is frozen; until then the derived lines recompute.
    is_snapshot: bool = False
    generated_at: ISTDateTime | None = None


class DmrEnteredIn(BaseModel):
    """Only the lines nothing else in the system observes."""

    on_road: int | None = Field(default=None, ge=0, le=10_000)
    spare: int | None = Field(default=None, ge=0, le=10_000)
    under_warranty: int | None = Field(default=None, ge=0, le=10_000)
    rto_passing: int | None = Field(default=None, ge=0, le=10_000)
    high_energy_consumption: int | None = Field(default=None, ge=0, le=10_000)
    deep_cleaning: int | None = Field(default=None, ge=0, le=10_000)
    washed_cleaned: int | None = Field(default=None, ge=0, le=10_000)
    depot_accidents: int | None = Field(default=None, ge=0, le=10_000)
    tyres_scrapped: int | None = Field(default=None, ge=0, le=10_000)
    body_damages: int | None = Field(default=None, ge=0, le=10_000)
    notes: str | None = None


class DmrMonthOut(BaseModel):
    """The month grid: parameters down, one column per day."""

    site_code: str
    month: str
    dates: list[date_t]
    lines: list[DmrLine]
    #: line key -> value per date, in `dates` order. Plain numbers: a grid of
    #: quoted decimals is a parsing trap for every consumer.
    values: dict[str, list[float | None]]


# --- breakdown investigation (Annexure-V) ----------------------------------


class InvestigationOut(BaseModel):
    entry_id: str
    registration_no: str
    model: str = ""
    odometer_km: int | None = None
    driver_id: str | None = None
    defect_type: str = ""
    breakdown_reason: str = ""
    location: str | None = None
    breakdown_time: HHMM | None = None
    mechanic_reported_time: HHMM | None = None
    attended_time: HHMM | None = None
    loss_km: DecimalOut = None
    attended_details: str | None = None
    entry_date: date_t

    #: The investigation itself.
    findings: str | None = None
    last_pm_on: date_t | None = None
    last_pm_findings: str | None = None
    related_complaints: str | None = None
    investigation_action: str | None = None
    is_complete: bool = False
    updated_by: str = ""


class InvestigationList(BaseModel):
    site_code: str
    report_date: date_t
    items: list[InvestigationOut]
    outstanding: int = 0
    #: The closest date that does have breakdowns, when this one has none. An
    #: empty pane on a quiet day is indistinguishable from a broken one without
    #: it. Null when the site has no breakdowns at all.
    nearest_date: date_t | None = None


class InvestigationIn(BaseModel):
    findings: str | None = None
    last_pm_on: date_t | None = None
    last_pm_findings: str | None = None
    related_complaints: str | None = None
    investigation_action: str | None = None


# --- off-road cases ---------------------------------------------------------


class OffRoadOut(BaseModel):
    id: str
    site_code: str
    vehicle_id: str
    registration_no: str
    model: str = ""
    odometer_km: int | None = None
    issue: str = ""
    action_taken: str | None = None
    category: DefectCategory
    off_road_since: date_t
    expected_days: int | None = None
    expected_ready_on: date_t | None = None
    returned_on: date_t | None = None
    spare_parts_required: str | None = None
    remarks: str | None = None
    awaiting_vendor: bool = False
    #: How long it has been down as at the date asked for.
    days_down: int = 0
    #: True once it has been down longer than the report's threshold.
    is_held: bool = False


class OffRoadList(BaseModel):
    site_code: str
    report_date: date_t
    items: list[OffRoadOut]


class OffRoadIn(BaseModel):
    vehicle_id: str = Field(min_length=1, max_length=32)
    entry_id: str | None = None
    issue: str = ""
    action_taken: str | None = None
    category: DefectCategory = DefectCategory.other
    off_road_since: date_t | None = None
    expected_days: int | None = Field(default=None, ge=0, le=365)
    expected_ready_on: date_t | None = None
    odometer_km: int | None = Field(default=None, ge=0, le=10_000_000)
    spare_parts_required: str | None = None
    remarks: str | None = None
    awaiting_vendor: bool = False


class OffRoadClose(BaseModel):
    returned_on: date_t


class ChartCellOut(BaseModel):
    """One block on a control chart."""

    #: Empty means the block is blank — no topping, no inspection, no complaint.
    value: str = ""
    #: What the depot colours it: plain, pm, docking, breakdown.
    mark: CellMark = CellMark.plain
    #: The full text when `value` was shortened to fit the block. Empty when
    #: the two are the same.
    title: str = ""


class ChartRowOut(BaseModel):
    vehicle_id: str
    registration_no: str
    #: One per date, in `dates` order, always the same length as `dates`.
    cells: list[ChartCellOut]


class ChartKindOut(BaseModel):
    """A chart offered by the site, whether or not it has data behind it."""

    kind: ChartKind
    title: str
    legend: str
    unit: str = ""
    #: False when nothing in the system can answer it; the client says why
    #: rather than drawing an empty grid.
    available: bool = True
    unavailable_reason: str = ""


class ControlChartOut(ChartKindOut):
    site_code: str
    from_date: date_t
    to_date: date_t
    dates: list[date_t]
    rows: list[ChartRowOut]
    #: How many blocks carry anything — a whole-chart "is this being kept up".
    filled: int = 0


# --- fitted units: the statement and the history card ------------------------


class UnitTypeOut(BaseModel):
    id: int
    name: str
    sort_order: int = 0
    #: Marks the unit the DMR counts under "HV batteries replaced".
    is_hv_battery: bool = False


class FittedUnitOut(BaseModel):
    """One component's stay on one bus."""

    id: str
    site_code: str
    vehicle_id: str
    registration_no: str
    unit_type_id: int
    unit_name: str
    #: The Work Done entry this was fit alongside, when it was fit that way.
    entry_id: str | None = None
    unit_no: str | None = None
    fitted_on: date_t
    fitted_odometer_km: int | None = None
    removed_on: date_t | None = None
    removed_odometer_km: int | None = None
    #: Removal odometer less fitting odometer. Null when either is missing — an
    #: unknown life is not a life of zero.
    kms_covered: int | None = None
    removal_reason: str | None = None
    remarks: str | None = None
    #: False once it has come off.
    is_fitted: bool = True


class FittedUnitList(BaseModel):
    site_code: str
    month: str = ""
    items: list[FittedUnitOut]


class FitUnitIn(BaseModel):
    vehicle_id: str = Field(min_length=1, max_length=32)
    unit_type_id: int
    fitted_on: date_t
    #: Set when this fit is recorded alongside a Work Done entry's own save —
    #: lets the register list find it without guessing from vehicle + date.
    entry_id: str | None = Field(default=None, max_length=32)
    unit_no: str | None = Field(default=None, max_length=120)
    #: Defaults to the bus's last odometer reading when left out.
    fitted_odometer_km: int | None = Field(default=None, ge=0, le=10_000_000)
    remarks: str | None = None


class RemoveUnitIn(BaseModel):
    removed_on: date_t
    removed_odometer_km: int | None = Field(default=None, ge=0, le=10_000_000)
    removal_reason: str | None = None
    remarks: str | None = None


class HistoryEventOut(BaseModel):
    """What happened to one unit in one month."""

    #: "fitted", "removed", or "replaced" when both happened that month.
    kind: str
    #: What the block shows — the day, or both days.
    label: str
    unit_no: str = ""
    reason: str = ""
    kms_covered: int | None = None


class HistoryRowOut(BaseModel):
    unit_type_id: int
    unit_name: str
    #: One per month in `months` order; null where nothing happened.
    cells: list[HistoryEventOut | None]
    #: True when the unit is on the bus as at the end of the window.
    fitted_now: bool = False


class BusHistoryOut(BaseModel):
    site_code: str
    vehicle_id: str
    registration_no: str
    #: `yyyy-MM`, oldest first.
    months: list[str]
    rows: list[HistoryRowOut]
    #: How many blocks carry anything — whether the card is being kept at all.
    events: int = 0
